t/mojo/reactor_poll.t

```diff
@@ -7,332 +7,374 @@ use IO::Socket::INET;
-    my $reactor = shift;
-    $reactor->remove($two) if ++$again > 3;
-  }
-);
-$reactor->timer(
-  0.025 => sub {
-    my $reactor = shift;
-    $reactor->again($one, 0.025);
-    $reactor->again($two, 0.025);
-  }
-);
-$reactor->start;
-ok $after, 'timer was triggered';
-ok(($after - $before) < 200, 'less than 200 seconds');
-is $again, 4, 'recurring timer triggered four times';
-
-# Restart inactive timer
-$id = $reactor->timer(0 => sub { });
-ok $reactor->remove($id), 'removed';
-eval { $reactor->again($id) };
-like $@, qr/Timer not active/, 'right error';
-
-# Change inactive I/O watcher
-ok !$reactor->remove($listen), 'not removed again';
-eval { $reactor->watch($listen, 1, 1) };
-like $@, qr!I/O watcher not active!, 'right error';
-
-# Error
-my $err;
-$reactor->unsubscribe('error')->on(
-  error => sub {
-    shift->stop;
-    $err = pop;
-  }
-);
-$reactor->timer(0 => sub { die "works!\n" });
-$reactor->start;
-like $err, qr/works!/, 'right error';
-
-# Reset events
-$reactor->on(error => sub { });
-ok $reactor->has_subscribers('error'), 'has subscribers';
-$reactor->reset;
-ok !$reactor->has_subscribers('error'), 'no subscribers';
-
-# Recursion
-$timer   = undef;
-$reactor = $reactor->new;
-$reactor->timer(0 => sub { ++$timer and shift->one_tick });
-$reactor->one_tick;
-is $timer, 1, 'timer was triggered once';
-
-# Detection
-is(Mojo::Reactor::Poll->detect, 'Mojo::Reactor::Poll', 'right class');
-
-# Reactor in control
-is ref Mojo::IOLoop->singleton->reactor, 'Mojo::Reactor::Poll', 'right object';
-ok !Mojo::IOLoop->is_running, 'loop is not running';
-my ($buffer, $server_err, $server_running, $client_err, $client_running);
-$id = Mojo::IOLoop->server(
-  {address => '127.0.0.1'} => sub {
-    my ($loop, $stream) = @_;
-    $stream->write('test' => sub { shift->write('321') });
-    $server_running = Mojo::IOLoop->is_running;
-    eval { Mojo::IOLoop->start };
-    $server_err = $@;
-  }
-);
-$port = Mojo::IOLoop->acceptor($id)->port;
-Mojo::IOLoop->client(
-  {port => $port} => sub {
-    my ($loop, $err, $stream) = @_;
-    $stream->on(
-      read => sub {
-        my ($stream, $chunk) = @_;
-        $buffer .= $chunk;
-        return unless $buffer eq 'test321';
-        Mojo::IOLoop->singleton->reactor->stop;
-      }
-    );
-    $client_running = Mojo::IOLoop->is_running;
-    eval { Mojo::IOLoop->start };
-    $client_err = $@;
+
+subtest "Listen" => sub {
+  $reactor->io($listen => sub { pop() ? $writable++ : $readable++ })->watch($listen, 0, 0)->watch($listen, 1, 1);
+  $reactor->timer(0.025 => sub { shift->stop });
+  $reactor->start;
+  ok !$readable, 'handle is not readable';
+  ok !$writable, 'handle is not writable';
+};
+
+my $client;
+subtest "Connect" => sub {
+  $client = IO::Socket::INET->new(PeerAddr => '127.0.0.1', PeerPort => $port);
+  $reactor->timer(1 => sub { shift->stop });
+  $reactor->start;
+  ok $readable, 'handle is readable';
+  ok !$writable, 'handle is not writable';
+
+};
+
+my $server;
```

Comment:
Variable outside of subtest.

Reply:
variable shared between tests and moved out of subtest.  How best to isolate?